### PR TITLE
Go 1.6

### PR DESCRIPTION
**Pull request by leonklingele**

### CODE REVIEW

File: Library/Formula/go.rb [hunk 1]
```diff
@@ -15,9 +15,9 @@ class Go < Formula
   end
 
   devel do
```
Comment (DomT4) [inline Library/Formula/go.rb]: If there's no current `devel` release just kill the block, Thanks!

Comment (leonklingele) [inline Library/Formula/go.rb]: Makes sense. Pushed.

**Comment by jingweno** [comment]
:+1:

**Comment by DomT4** [comment]
Our CI system got hung up on some long builds earlier tonight. This will be merged as soon as possible, regardless of thumb count :wink:.

**Comment by theckman** [comment]
@DomT4 Apologies if this question is inappropriate for this PR. Is there a publicly visible endpoint to view build status? Trying to debate on whether to install from official sources, or wait for the Homebrew build queue to flush.

**Comment by DomT4** [comment]
Build queue is public at http://bot.brew.sh/job/Homebrew%20Pull%20Requests/. It appears to be about 3 builds away (Although that's not a guarantee that it'll pass first time or be merged instantly after, etc etc). Shouldn't be _too_ long.

**Comment by lukanus** [comment]
+1

**Comment by theckman** [comment]
That build failure is troublesome. Might indicate that the build environment contains a folder that causes the go tooling to abort:

>Go 1.5 introduced experimental support for a “vendor” directory that was enabled by an environment variable. In Go 1.6, the feature is now enabled by default. Source trees that contain a directory named “vendor” that is not used in accordance with the new feature will require changes to avoid broken builds (the simplest fix is to rename the directory).

**Comment by CamJN** [comment]
In go 1.6 `./src/vendor` exists in the source tarball.

**Comment by DomT4** [comment]
The failure is with `godep` rather than `go`. It's not a build error, "just" a dep error. As far as I can see https://github.com/tools/godep/commit/955e36c90cb22b2315f48ab773b653532627d53d might have addressed this previously, but if Go 1.6 sets `GO15VENDOREXPERIMENT` unconditionally without an opt-out I guess that nullifies that.

**Comment by DomT4** [comment]
Ah, the workaround has moved to `ENV["GO15VENDOREXPERIMENT"] = "0"`.

**Comment by theckman** [comment]
Right. But should we employ a workaround that will break on future releases?

**Comment by DomT4** [comment]
Stuff that uses godep:

* fugu - Fails without `ENV["GO15VENDOREXPERIMENT"] = "0"` set.
```
==> godep go build -o /usr/local/Cellar/fugu/1.1.1/bin/fugu main.go usage.go version.go
package main
	imports github.com/mattes/fugu
	imports github.com/docker/docker/registry
	imports github.com/docker/docker/engine
	imports github.com/docker/docker/utils
	imports github.com/docker/docker/pkg/archive
	imports github.com/docker/docker/vendor/src/code.google.com/p/go/src/pkg/archive/tar: must be imported as src/code.google.com/p/go/src/pkg/archive/tar
godep: go exit status 1
```

* vultr - Works fine with the latest `godep` but with this warning, so probably wants the ENV:
```
==> godep go build -o vultr .
godep: WARNING: Go version (go1.6) & $GO15VENDOREXPERIMENT= wants to enable the vendor experiment, but disabling because a Godep workspace (Godeps/_workspace) exists
```

* doctl - Fails for me either way with:
```
==> godep restore
godep: Dep (github.com/digitalocean/godo) restored, but was unable to load it with error:
	Package (github.com/google/go-querystring/query) not found
godep: Dep (golang.org/x/oauth2) restored, but was unable to load it with error:
	Package (golang.org/x/net/context) not found
godep: Error checking some deps.
```

* cig - Works but wants the ENV because `godep` is working around the issue for 'em:
```
==> godep go build -o cig .
godep: WARNING: Go version (go1.6) & $GO15VENDOREXPERIMENT= wants to enable the vendor experiment, but disabling because a Godep workspace (Godeps/_workspace) exists
```

* forego is in the same boat:
```
==> godep go build -ldflags -X main.Version 0.16.1 -X main.allowUpdate false -o forego
godep: WARNING: Go version (go1.6) & $GO15VENDOREXPERIMENT= wants to enable the vendor experiment, but disabling because a Godep workspace (Godeps/_workspace) exists
# _/private/tmp/forego20160218-91660-6wl9ql/forego-0.16.1
link: warning: option -X main.Version 0.16.1 may not work in future releases; use -X main.Version=0.16.1
link: warning: option -X main.allowUpdate false may not work in future releases; use -X main.allowUpdate=false
```

* deis again:
```
==> godep go build -a -ldflags -s -o dist/deis
godep: WARNING: Go version (go1.6) & $GO15VENDOREXPERIMENT= wants to enable the vendor experiment, but disabling because a Godep workspace (Godeps/_workspace) exists
```

* docker-machine-parallels needs the new ENV set or it breaks.
```
==> godep go build -o ./bin/docker-machine-driver-parallels ./bin
# _/private/tmp/docker-machine-parallels20160218-92268-1bl0h9m/docker-machine-parallels-1.1.1/bin
bin/main.go:39: cannot use parallels.NewDriver("", "") (type *parallels.Driver) as type "github.com/docker/machine/libmachine/drivers".Driver in argument to plugin.RegisterDriver:
	*parallels.Driver does not implement "github.com/docker/machine/libmachine/drivers".Driver (wrong type for GetCreateFlags method)
		have GetCreateFlags() []"github.com/Parallels/docker-machine-parallels/vendor/github.com/docker/machine/libmachine/mcnflag".Flag
		want GetCreateFlags() []"github.com/docker/machine/libmachine/mcnflag".Flag
godep: go exit status 2
```

* deisctl is fine but needs the ENV because `godep` is doing the hack for 'em:
```
==> godep go build -a -ldflags -s -o dist/deisctl
godep: WARNING: Go version (go1.6) & $GO15VENDOREXPERIMENT= wants to enable the vendor experiment, but disabling because a Godep workspace (Godeps/_workspace) exists
```

* vault can't be tested right now because apparently they retagged the release which has broken our checksum mechanism there. That's nice.

* ipfs needs the new ENV or :skull: -
```
==> make build
go build -ldflags="-X "github.com/ipfs/go-ipfs/repo/config".CurrentCommit=7070b4d"
package .
	imports github.com/ipfs/go-ipfs/vendor/QmQg1J6vikuXF9oDvm4wpdeAUvvkVEKW1EYDw9HhTMnP2b/go-log: must be imported as QmQg1J6vikuXF9oDvm4wpdeAUvvkVEKW1EYDw9HhTMnP2b/go-log
make: *** [build] Error 1
```

* corectl is fine but needs the ENV to stop `godep` doing it silently:
```
==> make corectl VERSION=0.5.4
fatal: Not a git repository (or any of the parent directories): .git
godep go build -o corectl  -ldflags " -w -s -X main.Version=0.5.4 -X main.BuildDate=2016-02-18T02:45:45GMT"
godep: WARNING: Go version (go1.6) & $GO15VENDOREXPERIMENT= wants to enable the vendor experiment, but disabling because a Godep workspace (Godeps/_workspace) exists
```

* wellington is the same deal as the above.

* terraform fails to build either way for me locally.

**Comment by DomT4** [comment]
> Right. But should we employ a workaround that will break on future releases?

Depends. If by future release you mean 1.7.0, these formulae can definitely be tided over with the new `ENV` until upstream handles that themselves. If by future release you mean 1.6.1, that'll be trickier.

**Comment by DomT4** [comment]
First things first, I'll file a PR to update `godep` to the newest release so it can actually handle the ENV if we decide to go down this route.

**Comment by ewok** [comment]
I think there must be new hashs
  bottle do
    sha256 "XXX" => :el_capitan
    sha256 "XXX" => :yosemite
    sha256 "XXX" => :mavericks
  end

**Comment by mattfarina** [comment]
Can someone explain why the install of Go is dependency on godep? Shouldn't it be the other way around. How can we get the godep test removed from the install of Go?

**Comment by joefitzgerald** [comment]
> Can someone explain why the install of Go is dependency on godep? Shouldn't it be the other way around. How can we get the godep test removed from the install of Go?

:1: to that.

**Comment by brettlangdon** [comment]
@mattfarina @joefitzgerald it isn't that the `go` formula depends on `godep`, it is the other way around. `godep` is having issues with `go1.6` that needs to get resolved.

There was #49290 which aims to resolve the issue.

**Comment by joefitzgerald** [comment]
@brettlangdon But why would a godep issue associated with other applications block the release of the go 1.6 formula? Seems as though the dependency tree is inverted. The maintainers of the aforementioned formula should have been tracking go 1.6 via the devel releases and updated accordingly, prior to now. And the CI for *those* packages should have failed when the devel formula for go 1.6 was released (i.e. the betas) so they were notified.

I totally understand that this is maintained by volunteers (I'm a volunteer too https://github.com/joefitzgerald) so I'm not trying to act like you owe users anything; I'm just trying to assert that the CI dependency tree could be improved / inverted such that dependencies break when go releases a new version, but go doesn't break if it is otherwise valid.

**Comment by damienstanton** [comment]
Looking to assist but I am confused - why is `godep` being tested as a part of the `go` release formula?

And why is `GO15VENDOREXPERIMENT` still enabled on the build server when it is implicitly implemented as of 1.6?

**Comment by mikemcquaid** [comment]
> I'm just trying to assert that the CI dependency tree could be improved / inverted such that dependencies break when go releases a new version, but go doesn't break if it is otherwise valid.

This isn't how package managers work. We need to ensure that we don't break other software when we update some.

Please everyone be patient.

**Comment by DomT4** [comment]
Folks, can we please keep it together. I was up till gone 5am looking at this, I had to go sleep at _some point_. I'm not making you all listen to Taylor Swift on repeat; I'm just asking for some patience here. We are working through this as quickly as we can.

**Comment by CamJN** [comment]
Sorry if I came across as impatient, just trying to help.

**Comment by damienstanton** [comment]
Same as @CamJN; I just want to help understand the problem. If the best I can do is sit patiently, I will do so gladly :+1:

**Comment by mattfarina** [comment]
@mikemcquaid Godep is just one of the package managers for Go. Others in homebrew aren't tested along with Go (and work with 1.6). A release of Godep from a week ago is in homebrew. Can you share more about the coupling here?

**Comment by mikemcquaid** [comment]
@mattfarina We automatically test reverse dependencies to ensure if we upgrade the package they won't be broken. In this case: `godep`'s formula would be broken as-is. @DomT4 is working on this. Any other Go pull request submitted to Homebrew will always find this broken. Our desire for `godep` users to not have things be broken is more important than rushing Go 1.6 in here.

**Comment by DomT4** [comment]
> I think there must be new hashes

Thanks for the suggestion but the `bottle do` block is fine to ignore on all and any updates; we have logic in place to replace that and it isn't used for any formulae modified inside the PR.

> godep is having issues with go1.6 that needs to get resolved.

This is closer to what's happening, but the "blame" doesn't really fall on `godep` here either. `godep` in the latest release introduced some magic that attempts to mitigate this by being clever about disabling the ENV automatically when necessary, but there's only so much they can do and realistically upstreams are where the core problem should/will be solved eventually.

Go previously introduced `GO15VENDOREXPERIMENT` but it had to be enabled; many formulae authors and upstreams started moving across to support that but naturally as we have some older packages and some upstreams that aren't as active it wasn't applied to everything.

The problem now is that it's on by default, the formulae that do not yet support the experiment have to be explicitly told to turn it off or the compile process breaks as bad as Walter White. This is the sort of thing that would have been _very quickly_ discovered by users if we'd ignored the `godep` test breaking and merged this.

So on our side, we now have to transition from formulae needing to be told to turn it on and formulae now needing to be told to turn it _off_. This involves a fair bit of local testing to know what's okay and what needs modifying, which I will work through as quickly as possible whilst attempting to not spend my whole day with the Go family.

> The maintainers of the aforementioned formula

We don't have individual formulae maintainers. If you like a formula and you stick with it? Great, we :heart: that kind of dedication. But most of our contributions are from casual contributors who help a bit here and there without specific focus.

> why is godep being tested as a part of the go release formula?

As Mike said the CI runs `brew uses --recursive --skip-build --skip-optional go` which produces a list of other formulae with **hard** `go` dependencies to check. Those formulae then have their bottles poured and `brew test --verbose` executed on them. This is useful to check for things like `dylib` version changes, API breaks, etc that we need to rebuild other packages to fix.

> And why is GO15VENDOREXPERIMENT still enabled on the build server when it is implicitly implemented as of 1.6?

It's being set in individual formulae known to support the option, as explained above. It's not globally applied.

**Comment by joefitzgerald** [comment]
@DomT4 @mikemcquaid thanks for the responses. This clarifies things for me - much appreciated.

The only item I don't understand is why this wasn't caught earlier when the 1.6 betas were shipping? Is `--devel` not part of the CI pipeline?

**Comment by mikemcquaid** [comment]
@joefitzgerald It is not. That is something we could do but we'd need more CI infrastructure (and therefore money) to be able to do that. We'll be able to get corporate donations in the near future more easily and I'll try to remember to reach out to Pivotal to ask for some :wink:

**Comment by bcomnes** [comment]
Just so that people are aware, homebrew is a distributed package manager based on git.  Nothing stopping you from the following:

```
$ cd /usr/local
$ git remote add go git@github.com:leonklingele/homebrew.git
$ git fetch go
$ git checkout go-1.6
$ brew upgrade go
$ git checkout master
$ git remote rm go
$ git branch -d go-1.6
```

**Comment by DomT4** [comment]
Merged in https://github.com/Homebrew/homebrew/commit/716abfd72fe673679b08902562453855117507cb. Thank you for your contribution to Homebrew @leonklingele; we appreciate it! :smiley_cat:.

@ Everyone; If you spot any issues I've missed here please report them, or ideally, file a PR to fix 'em. Major updates are difficult on package managers - community & upstream support massively helps resolve things quickly.

**Pull request closed without merging**